### PR TITLE
Add EpochService2 intial code

### PR DESCRIPTION
**Pull request by sydhds**
This is rewrite of the EpochService service to remove the need to specify a valid epoch slice to make it work properly. In the process, the code has been cleaned up and should be easier to read than the previous one.
Also unit tests have been improved.

### CODE REVIEW

**Comment by nadeemb53** [comment]
pr needs a description

**Comment by sydhds** [comment]
> pr needs a description

Would you like to review this PR now @nadeemb53 ?

File: rln-prover/prover/src/epoch_service_2.rs [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+const WAIT_UNTIL_MIN_DURATION: Duration = Duration::from_secs(10);
+/// EpochService::compute_wait_until() can return an error like TooLow (see WAIT_UNTIL_MIN_DURATION)
+/// so the epoch service will retry X many times.
+const WAIT_UNTIL_MAX_RETRY: usize = 5;
+
+pub struct EpochService2 {
+    /// Duration of an epoch (quotas reset every epoch)
+    epoch_duration: Duration,
+    /// Current epoch
+    pub current_epoch: Arc<RwLock<Epoch>>,
```
Comment (nadeemb53) [inline rln-prover/prover/src/epoch_service_2.rs]: This and `epoch_changes` below are public, which leaks the internal locking strategy and lets callers hold the `RwLock` write guard across an `.await` and deadlock the service. Make it `pub(crate)` and expose accessors:            
                                                                                                                                                                                                                             
  ```rust                                                                                                                                                                                                                                
  pub(crate) current_epoch: Arc<RwLock<Epoch>>,                   
  Then add pub fn current_epoch(&self) -> Epoch { *self.current_epoch.read() }. 
```

File: rln-prover/prover/src/epoch_service_2.rs [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+    /// Genesis time (aka when the service has been started at the first time)
+    genesis: DateTime<Utc>,
+    /// Channel to notify when an epoch has just changed
+    pub epoch_changes: Arc<Notify>,
```
Comment (nadeemb53) [inline rln-prover/prover/src/epoch_service_2.rs]: Same as above, make it pub(crate) and expose `pub fn subscribe(&self) -> Arc<Notify> { self.epoch_changes.clone() }`. 

Also: `Notify::notify_one()` (used at line 110) only wakes a single waiter. Right now `UserDbService::listen_for_epoch_changes` is the only consumer, but if a second subscriber is ever added (metrics exporter, slasher reaction, etc.), one of them will silently miss epoch transitions. Consider switching the entire mechanism to `tokio::sync::watch::<Epoch>` , that's a better fit for `"current value + change notification"` anyway.

File: rln-prover/prover/src/epoch_service_2.rs [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+// Internal
+use crate::error::AppError2;
+use crate::metrics::{
+    EPOCH_SERVICE_CURRENT_EPOCH, EPOCH_SERVICE_CURRENT_EPOCH_SLICE, EPOCH_SERVICE_DRIFT_MILLIS,
+};
+
+/// EpochService2::compute_wait_until() can return a low duration (see Notes in listen_for_new_epoch)
+/// So we need a minimum threshold to accept or not this duration
+const WAIT_UNTIL_MIN_DURATION: Duration = Duration::from_secs(10);
+/// EpochService::compute_wait_until() can return an error like TooLow (see WAIT_UNTIL_MIN_DURATION)
```
Comment (nadeemb53) [inline rln-prover/prover/src/epoch_service_2.rs:19]: 10s is too coarse for short epochs. Test/local config uses `epoch_duration_secs = 60`. If the prover starts in the last 10 s of an epoch, `compute_wait_until` returns `TooLow`, the loop sleeps 10 s, retries, and with `WAIT_UNTIL_MAX_RETRY = 5` (line 21) it can spend up to 50 s stuck in the retry loop on startup, most of an entire test epoch. e2e tests will be flaky for "the prover takes ages to start".

Either move both constants into `EpochService2Config` so they can be tuned per environment, or auto-scale:
  ```rust                                                         
  let min_wait = std::cmp::min(Duration::from_secs(10), self.epoch_duration / 6);
```

File: rln-prover/prover/src/epoch_service_2.rs [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+            match EpochService2::compute_wait_until(
+                &self.genesis,
+                &self.epoch_duration,
+                &|| Utc::now(),
+                &|| tokio::time::Instant::now(),
+            ) {
+                Ok((current_epoch, wait_until)) => break (current_epoch, wait_until),
+                Err(WaitUntilError::TooLow(d1, d2)) => {
+                    // Wait until is too low (according to const WAIT_UNTIL_MIN_DURATION)
+                    // so we will retry (WAIT_UNTIL_MAX_COMPUTE_ERROR many times) after a short sleep
```
Comment (nadeemb53) [inline rln-prover/prover/src/epoch_service_2.rs]: isn't the constant WAIT_UNTIL_MAX_RETRY, not WAIT_UNTIL_MAX_COMPUTE_ERROR?

File: rln-prover/prover/src/epoch_service_2.rs [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+                Err(e) => {
+                    // Another error (like OutOfRange) - exiting...
+                    error!("Error computing the initial wait until: {}", e);
+                    return Err(AppError2::EpochError(e));
+                }
+            }
+        };
+        // Debug
+        // let current_epoch = 0;
+        // let mut wait_until = tokio::time::Instant::now();
```
Comment (nadeemb53) [inline rln-prover/prover/src/epoch_service_2.rs]: Stale debug scaffolding

File: rln-prover/prover/src/epoch_service_2.rs [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+
+        *self.current_epoch.write() = current_epoch.into();
+        debug!("Initial epoch: {}", current_epoch);
+
+        loop {
+            debug!("wait until: {:?}", wait_until);
+            // XXX: Should we check the drift between now() and wait_until ?
```
Comment (nadeemb53) [inline rln-prover/prover/src/epoch_service_2.rs:77]: The histogram!(EPOCH_SERVICE_DRIFT_MILLIS) call right below already does this, no?

File: rln-prover/prover/src/epoch_service_2.rs [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+
+                current_epoch += 1;
+                *self.current_epoch.write() = current_epoch.into();
+
+                // Note: based on this link https://doc.rust-lang.org/reference/expressions/operator-expr.html#type-cast-expressions
+                //       "Casting from an integer to float will produce the closest possible float *"
+                gauge!(EPOCH_SERVICE_CURRENT_EPOCH.name, "prover" => "epoch service")
+                    .set(current_epoch as f64);
+
+                self.epoch_changes.notify_one();
```
Comment (nadeemb53) [inline rln-prover/prover/src/epoch_service_2.rs:108]: See comment on line 31, notify_one() will silently starve future subscribers. If you don't want to refactor to tokio::sync::watch in this PR, add a comment here documenting the single-subscriber assumption.

File: rln-prover/prover/src/epoch_service_2.rs [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+            }
+        }
+    }
+
+    fn compute_wait_until<T, F: Fn() -> DateTime<Utc>, TF: Fn() -> T>(
+        genesis: &DateTime<Utc>,
+        epoch_duration: &Duration,
+        now: &F,
+        now2: &TF,
```
Comment (nadeemb53) [inline rln-prover/prover/src/epoch_service_2.rs]: now and now2 are wall-clock vs monotonic-clock fakes for testability. The names tell you nothing. Rename for clarity:
```rust
  fn compute_wait_until<T, F: Fn() -> DateTime<Utc>, TF: Fn() -> T>(
      genesis: &DateTime<Utc>,
      epoch_duration: &Duration,
      wall_clock_now: &F,
      monotonic_now: &TF,
  ) -> Result<(i64, T), WaitUntilError>
```

File: rln-prover/prover/src/epoch_service_2.rs [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+        //       but in unit test, we use now() == now2() as it is easier to manipulate DateTime<Utc>
+
+        let (current_epoch, _current_epoch_start_ts, current_epoch_end_ts) =
+            Self::compute_current_epoch_0(genesis, now, epoch_duration);
+
+        // time to wait to next epoch
+        let now_ = now();
+        let now_2_ = now2();
+        let wait_until_0_ = current_epoch_end_ts - now_.timestamp();
+        let wait_until_0 = Duration::from_secs(wait_until_0_ as u64);
```
Comment (nadeemb53) [inline rln-prover/prover/src/epoch_service_2.rs]: Sub-second precision is lost here:
```
  let wait_until_0_ = current_epoch_end_ts - now_.timestamp();   // both i64 SECONDS
  let wait_until_0 = Duration::from_secs(wait_until_0_ as u64);
```
 now_.timestamp() returns whole seconds, so this throws away the milliseconds part of now_. The prover wakes up up to 999 ms BEFORE the actual epoch boundary every cycle, and the `EPOCH_SERVICE_DRIFT_MILLIS` histogram will look noisy because of it.

 Worse: if wait_until_0_ is ever negative (system clock skew, NTP correction, suspend/resume), the as u64 cast wraps to ~`u64::MAX, and Duration::from_secs(...)`builds a Duration of ~584 billion years. The `nextsleep_until` would effectively never wake. 

  Fix:
```
  let wait_until_0_ms = current_epoch_end_ts.saturating_mul(1000)
      - now_.timestamp_millis();
  if wait_until_0_ms < 0 {
      return Err(WaitUntilError::TooLow(Duration::ZERO, WAIT_UNTIL_MIN_DURATION));
  }
  let wait_until_0 = Duration::from_millis(wait_until_0_ms as u64);
```
  The TooLow retry path then correctly recomputes for the next epoch.

File: rln-prover/prover/src/epoch_service_2.rs [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+
+    /// Compute current epoch since genesis
+    /// Return current_epoch AND DateTime of the start & end of this current epoch (as timestamps)
+    fn compute_current_epoch_0<F: Fn() -> DateTime<Utc>>(
+        genesis: &DateTime<Utc>,
+        now: &F,
+        epoch_duration: &Duration,
+    ) -> (i64, i64, i64) {
+        debug_assert!(now() >= *genesis);
+        debug_assert!(i64::try_from(epoch_duration.as_secs()).is_ok());
```
Comment (nadeemb53) [inline rln-prover/prover/src/epoch_service_2.rs:161]: debug_assert! is a no-op in release builds. If now() < *genesis (clock skew, container started before genesis time, NTP rewind), diff = now_timestamp - genesis_timestamp is negative, current_epoch =
  diff.checked_div(epoch_dur).unwrap() becomes a negative i64, and that propagates as the "current epoch", every downstream Epoch::from(current_epoch) will be wrong.

Replace with a hard check that returns an error:
```
  if now() < *genesis {
      return /* Err(...) */;
  }
```
You'll need a new WaitUntilError variant (or surface this via the existing EpochServiceInitError::FutureGenesis)

File: rln-prover/prover/src/epoch_service_2.rs [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+            genesis,
+        }
+    }
+}
+
+impl TryFrom<EpochService2Config> for EpochService2 {
+    type Error = EpochServiceInitError;
+
+    fn try_from(config: EpochService2Config) -> Result<Self, Self::Error> {
+        if config.genesis >= Utc::now() {
```
Comment (nadeemb53) [inline rln-prover/prover/src/epoch_service_2.rs:203]: Off-by-one. The error is FutureGenesis ("genesis is in the future") but >= rejects genesis == now, which is technically valid (zero seconds in the past). Use strict >:
```
  if config.genesis > Utc::now() {
```
  Edge case but it'll break a unit test that uses Utc::now() as genesis.

File: rln-prover/prover/src/epoch_service_2.rs [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+            genesis: config.genesis,
+            epoch_changes: Arc::new(Default::default()),
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use alloy::providers::WatchTxError::Timeout;
```
Comment (nadeemb53) [inline rln-prover/prover/src/epoch_service_2.rs]: Timeout is never referenced in the test module

File: rln-prover/prover/src/epoch_service_2.rs [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+            epoch_duration: config.epoch_duration,
+            current_epoch: Arc::new(Default::default()),
+            genesis: config.genesis,
+            epoch_changes: Arc::new(Default::default()),
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (nadeemb53) [inline rln-prover/prover/src/epoch_service_2.rs:223]: the unit tests cover compute_current_epoch_0 and compute_wait_until thoroughly, but the actual listen_for_new_epoch async loop has no test. That's the most behavior-rich part of the new file:

  - The retry loop on WaitUntilError::TooLow (max 5 retries, 10 s sleep)
  - The EpochServiceOverflow branch
  - The notify_one() interaction with UserDbService::listen_for_epoch_changes

spawn listen_for_new_epoch with a 100 ms epoch duration, wait for 3 ticks via the Notify, assert current_epoch advanced exactly 3 times. Same kind of mocked-time setup you already have for the pure functions.

File: rln-prover/prover/src/lib.rs [hunk 14]
```diff
@@ -94,6 +96,17 @@ pub async fn run_prover(app_args: AppArgs) -> Result<(), AppError2> {
...
+        "Starting epoch service: epoch_duration={}s",
+        app_args.epoch_duration_secs
+    );
+    let epoch_config = EpochService2Config::new(epoch_duration, ARGS_DEFAULT_GENESIS);
+
+    let epoch_service =
+        EpochService2::try_from(epoch_config).expect("Failed to create epoch service");
 
     // Alloy provider (Smart contract provider)
     let provider = if app_args.ws_rpc_url.is_some() {
```
Comment (nadeemb53) [inline rln-prover/prover/src/lib.rs]: I see that the entire old EpochService::try_from(epoch_config) block in a /* … */ comment. Stale commented-out code rots fast and the next reader will not know whether it's intentional or a partial revert.

  Either:
  - (a) Delete the block here AND delete epoch_service.rs entirely (move the still-used types: Epoch, EpochServiceInitError, WaitUntilError, DEFAULT_EPOCH_DURATION  into a small epoch_common.rs module that epoch_service_2.rs imports), OR
  - (b) If you want a deprecation period, mark EpochService with #[deprecated], document the removal milestone in a code comment, and still delete this block here (it's dead either way).

  Strongly prefer (a). The whole point of the PR is to simplify; leaving both versions in place undermines that.

  While you're at it, please rename epoch_service_2.rs → epoch_service.rs and EpochService2/EpochService2Config → EpochService/EpochServiceConfig. The "_2" suffix mirrors the existing user_db_2.rs cruft, and that suffix never gets cleaned up later, it just becomes permanent

File: rln-prover/prover/src/epoch_service_2.rs (hunk 3 above)
Comment (sydhds) [reply rln-prover/prover/src/epoch_service_2.rs:19]: Not sure about the reasoning here:

1- If the prover starts in the last 10s of an epoch, compute_wait_until return TooLow then waits for 10s so next call to compute_wait_until will NOT return TooLow

2- Can you explain the reason for usnig 6 as a divisor for epoch_duration because as you mentioned for your tests the epoch_duration is 60s so min_wait would still be 10s here no?

The only issue I see with this constant is using very low epoch_duration (around 12s or even less) but this makes little sense. I suggest we refuse epoch_duration below 30s here. wdyt?

File: rln-prover/prover/src/epoch_service_2.rs (hunk 4 above)
Comment (sydhds) [reply rln-prover/prover/src/epoch_service_2.rs]: Fixed

File: rln-prover/prover/src/epoch_service_2.rs (hunk 1 above)
Comment (sydhds) [reply rln-prover/prover/src/epoch_service_2.rs]: Using pub(crate) here...

File: rln-prover/prover/src/epoch_service_2.rs (hunk 2 above)
Comment (sydhds) [reply rln-prover/prover/src/epoch_service_2.rs]: Using pub(crate) here...

File: rln-prover/prover/src/epoch_service_2.rs (hunk 5 above)
Comment (sydhds) [reply rln-prover/prover/src/epoch_service_2.rs]: Removed

File: rln-prover/prover/src/epoch_service_2.rs (hunk 6 above)
Comment (sydhds) [reply rln-prover/prover/src/epoch_service_2.rs:77]: In dev, tracing is nice to have while in production RUST_LOG will likely set > debug and then the telemetry will be used. I can remove this debug! print if you want?

File: rln-prover/prover/src/epoch_service_2.rs (hunk 7 above)
Comment (sydhds) [reply rln-prover/prover/src/epoch_service_2.rs:108]: Doc added

File: rln-prover/prover/src/epoch_service_2.rs (hunk 8 above)
Comment (sydhds) [reply rln-prover/prover/src/epoch_service_2.rs]: Changed the name as suggested

File: rln-prover/prover/src/epoch_service_2.rs (hunk 10 above)
Comment (sydhds) [reply rln-prover/prover/src/epoch_service_2.rs:161]: The checks here are very unlikely to occur in prod mode and are just here to warn when dev'ing the unit tests...  Note this is already tested in `impl TryFrom<EpochService2Config> for EpochService2` impl

File: rln-prover/prover/src/epoch_service_2.rs (hunk 11 above)
Comment (sydhds) [reply rln-prover/prover/src/epoch_service_2.rs:203]: No it does not make sense to have the genesis and now() return the same timestamp here...

File: rln-prover/prover/src/epoch_service_2.rs (hunk 12 above)
Comment (sydhds) [reply rln-prover/prover/src/epoch_service_2.rs]: Removed unused

File: rln-prover/prover/src/lib.rs (hunk 14 above)
Comment (sydhds) [reply rln-prover/prover/src/lib.rs]: Option a. done

File: rln-prover/prover/src/epoch_service_2.rs (hunk 13 above)
Comment (sydhds) [reply rln-prover/prover/src/epoch_service_2.rs:223]: Added unit test: test_service

File: rln-prover/prover/src/epoch_service_2.rs (hunk 9 above)
Comment (sydhds) [reply rln-prover/prover/src/epoch_service_2.rs]: Done
